File: elixir/lib/design_patterns/abstract_factory.ex
defmodule WeaponFactory do
  @moduledoc """
  Abstract factory for creating weapons
  """
  @type t :: module()

  @callback create_sword() :: Sword.t()
  @callback create_gun() :: Gun.t()
end

defmodule Sword do
  @moduledoc """
  Abstract product for sword
  """
  @type t :: module()

  @callback swing() :: :ok
end

defmodule Gun do
  @moduledoc """
  Abstract product for gun
  """
  @type t :: module()

  @callback shoot() :: :ok
end

defmodule TexasWeaponFactory do
  @moduledoc """
  Concreate factory for creating weapons in Texas
  """
  @behaviour WeaponFactory

  def create_sword() do
    TexasSword
  end

  def create_gun() do
    TexasGun
  end
end

defmodule TexasSword do
  @moduledoc """
  Concrete product for sword in Texas
  """
  @behaviour Sword

  def swing() do
    IO.puts("[ Sharp metalic sound ]")
  end
end

defmodule TexasGun do
  @moduledoc """
  Concrete product for gun in Texas
  """
  @behaviour Gun

  def shoot() do
    IO.puts("[ Bbaak Bbaak ]")
  end
end

defmodule ChildWeaponFactory do
  @moduledoc """
  Concreate factory for creating child's weapons
  """
  @behaviour WeaponFactory

  def create_sword() do
    ChildSword
  end

  def create_gun() do
    ChildGun
  end
end

defmodule ChildSword do
  @moduledoc """
  Concrete product for child's sword
  """
  @behaviour Sword

  def swing() do
    IO.puts("[ Plastic dull sound ]")
  end
end

defmodule ChildGun do
  @moduledoc """
  Concrete product for child's gun
  """
  @behaviour Gun

  def shoot() do
    IO.puts("[ Nerf Nerf ]")
  end
end

defmodule WeaponClient do
  @spec create_and_test_weapons(factory :: WeaponFactory.t()) :: :ok
  def create_and_test_weapons(factory) do
    sword = factory.create_sword()
    sword.swing()

    gun = factory.create_gun()
    gun.shoot()
  end
end

File: rust/src/abstract_factory.rs
mod abstract_definitions {
    // Abstract Products
    pub trait Sword {
        fn swing(&self);
    }

    pub trait Gun {
        fn shoot(&self);
    }

    // Abstract Factory (w/ associated types)
    pub trait WeaponFactory {
        type S: Sword;
        type G: Gun;

        fn create_sword(&self) -> Self::S;
        fn create_gun(&self) -> Self::G;
    }

    // Abstract Factory (w/ box pointer)
    pub trait DynWeaponFactory {
        fn create_sword(&self) -> Box<dyn Sword>;
        fn create_gun(&self) -> Box<dyn Gun>;
    }
}

mod texas {
    use super::abstract_definitions::{Gun, Sword, WeaponFactory};

    // Concrete Product
    pub struct TexasSword;

    impl Sword for TexasSword {
        fn swing(&self) {
            println!("[ Sharp swinging sound ]")
        }
    }

    pub struct TexasGun;

    impl Gun for TexasGun {
        fn shoot(&self) {
            println!("[ Bbaak Bbaak ]")
        }
    }

    // Concrete Factory
    pub struct TexasWeaponFactory;

    impl WeaponFactory for TexasWeaponFactory {
        type S = TexasSword;
        type G = TexasGun;

        fn create_sword(&self) -> Self::S {
            TexasSword
        }

        fn create_gun(&self) -> Self::G {
            TexasGun
        }
    }
}

mod child {
    use super::abstract_definitions::{Gun, Sword, WeaponFactory};

    // Concrete Product
    pub struct ChildSword;

    impl Sword for ChildSword {
        fn swing(&self) {
            println!("[ Plastic sound ]")
        }
    }

    pub struct ChildGun;

    impl Gun for ChildGun {
        fn shoot(&self) {
            println!("[ Nerf Nerf ]")
        }
    }

    // Concrete factory
    pub struct ChildWeaponFactory;

    impl WeaponFactory for ChildWeaponFactory {
        type S = ChildSword;
        type G = ChildGun;

        fn create_sword(&self) -> Self::S {
            ChildSword
        }

        fn create_gun(&self) -> Self::G {
            ChildGun
        }
    }
}

mod client {
    use super::abstract_definitions::{Gun, Sword, WeaponFactory};

    // Client code
    pub fn create_weapons(factory: impl WeaponFactory) {
        let sword = factory.create_sword();
        sword.swing();

        let gun = factory.create_gun();
        gun.shoot();
    }
}
